src/common.cc

```diff
@@ -130,20 +132,27 @@ double get_time() {
   return tv.tv_sec + (tv.tv_nsec * 1e-9);
 }
 
-/* Converts '~/...' paths to '/home/blah/...'.  It's similar to
- * variable_substitute, works for any enviroment variable */
-std::string to_real_path(const std::string &source) {
-    wordexp_t p;
-    char **w;
-    int i;
-    const char *csource = source.c_str();
-    if (wordexp(csource, &p, 0) != 0) {
-        return nullptr;
+/// @brief Handles environment variable expansion in paths and canonicalization.
+///
+/// Examples:
+/// - `~/conky` -> `/home/conky_user/conky`
+/// - `$HOME/conky` -> `/home/conky_user/conky`
+/// - `$HOME/a/b/../c/../../conky` -> `/home/conky_user/conky`
+std::filesystem::path to_real_path(const std::filesystem::path &source) {
+  try {
+    std::string input = source.string();
+    if (source.string().compare(0, 1, "~") == 0) {
+      const char *home = std::getenv("HOME");
+      if (home) { input = std::string(home) + source.string().substr(1); }
     }
-    w = p.we_wordv;
-    const char *resolved_path = strdup(w[0]);
-    wordfree(&p);
-    return std::string(resolved_path);
+    std::string expanded = variable_substitute(input);
+    std::filesystem::path absolute = std::filesystem::absolute(expanded);
+    return std::filesystem::weakly_canonical(absolute);
+  } catch (const std::filesystem::filesystem_error &e) {
+    // file not found or permission issues
+    NORM_ERR("can't canonicalize path: %s", source.c_str());
+    return source;
+  }
```

Comment:
@brndnmtthws I requested review mostly for this change that affects non-BSD. I'd merge myself if it weren't for that (given that rest of changes won't "build less" haha). Because I hijacked g0mb4's PR, GitHub allows me to merge without a review, so I'll wait for you to merge it as intended.

Reply:
I see possible need for: `~USERNAME`, command substitution, maaaybe arithmetic expansion, limited wildcard expansion (producing a single-match path, fail on many)

Other are a side effect of how POSIX basically passes the content into shell. Arithmetic expansion is complicated because we need some way of recursively parsing it.

OpenBSD people say it should be replaced with `glob` if possible (makes sense for single-match thing described above). MacOS is deprecating it (it even used perl for implementation at some point). Windows has it's own thing (`CommandLineToArgvW`).

Whole glob thing also shows a bug with previous implementation - `to_real_path` could've returned a single string which contained multiple paths.

Should I make the current code a fallback for OpenBSD and use the old code or continue to implement a limited version of `wordexp`?

Reply:
Reverted old code for non-OpenBSD. Deferring this for a separate PR.